infrastructure/src/test/java/org/corfudb/infrastructure/CorfuSequencerServerTest.java

```diff
@@ -0,0 +1,528 @@
+        );
+
+        ArgumentCaptor<ResponseMsg> responseCaptor = ArgumentCaptor.forClass(ResponseMsg.class);
+        sequencerServer.handleMessage(request, mockChannelHandlerContext, mockServerRouter);
+
+        verify(mockServerRouter).sendResponse(responseCaptor.capture(), any(ChannelHandlerContext.class));
+        ResponseMsg response = responseCaptor.getValue();
+
+        // Assert that the payload has a SERVER_ERROR and that the base
+        // header fields have remained the same
+        assertTrue(compareBaseHeaderFields(request.getHeader(), response.getHeader()));
+        assertTrue(response.getPayload().hasServerError());
+        assertTrue(response.getPayload().getServerError().hasNotReadyError());
+    }
+
+
+    /**
+     * Test that if sequencer server's epoch is Layout.INVALID_EPOCH (startup) OR
+     * if epoch sent by the client is not the consecutive epoch of the sequencerEpoch
+     * then the sequencer should not accept bootstrapWithoutTailsUpdate BootstrapSequencerRequestMsg,
+     * and return a Sequencer.BootstrapSequencerResponseMsg.NACK message.
+     */
+    @Test
+    public void testRequireFullBootstrap() {
+        sequencerServer = new SequencerServer(mockServerContext, spyFactoryHelper);
+        sequencerServer.setSequencerEpoch(0L);
+
+        RequestMsg request = getRequestMsg(
+                getBasicHeader(false, true),
+                CorfuProtocolSequencer.getBootstrapSequencerRequestMsg(
+                        Collections.emptyMap(),
+                        0,
+                        0,
+                        true)
+        );
+
+        // Invoke the handler method
+        sequencerServer.handleMessage(request, mockChannelHandlerContext, mockServerRouter);
+
+        // Verify that sendResponse() was called and capture the response object
+        ArgumentCaptor<ResponseMsg> responseCaptor = ArgumentCaptor.forClass(ResponseMsg.class);
+        verify(mockServerRouter)
+                .sendResponse(responseCaptor.capture(), any(ChannelHandlerContext.class));
+        ResponseMsg response = responseCaptor.getValue();
+
+        // Assert that the payload has a BootstrapSequencerResponseMsg.NACK message
+        // and that the base header fields have remained the same
+        assertThat(compareBaseHeaderFields(request.getHeader(), response.getHeader())).isTrue();
+        assertThat(response.getPayload().hasBootstrapSequencerResponse()).isTrue();
+        assertThat(response.getPayload().getBootstrapSequencerResponse().getRespType())
+                .isEqualTo(Sequencer.BootstrapSequencerResponseMsg.Type.NACK);
+    }
+
+    /**
+     * Test that stale bootstrap requests are discarded.
+     * If epoch sent by the client is not greater than the sequencerEpoch then the sequencer should discard the
+     * BootstrapSequencerRequestMsg, and return a Sequencer.BootstrapSequencerResponseMsg.NACK message.
+     */
+    @Test
+    public void testStaleBootstrapRequest() {
```

Comment:
Can we also verify that some of the state was not updated? Methods like `setSequencerEpoch` and others should not have been invoked when the bootstrap request is stale.